examples/petsc/dmswarm.c

```diff
@@ -0,0 +1,990 @@
+// Copyright (c) 2017-2022, Lawrence Livermore National Security, LLC and other CEED contributors.
+// All Rights Reserved. See the top-level LICENSE and NOTICE files for details.
+//
+// SPDX-License-Identifier: BSD-2-Clause
+//
+// This file is part of CEED:  http://github.com/ceed
+
+//                        libCEED + PETSc DMSwarm Example
+//
+// This example demonstrates a simple usage of libCEED with DMSwarm.
+// This example combines elements of PETSc src/impls/dm/swam/tutorials/ex1.c and src/impls/dm/swarm/tests/ex6.c
+//
+// Build with:
+//
+//     make dmswarm [PETSC_DIR=</path/to/petsc>] [CEED_DIR=</path/to/libceed>]
+//
+// Sample runs:
+//
+//     ./dmswarm
+///
+//TESTARGS -ceed {ceed_resource} -test -dm_plex_dim 3 -dm_plex_box_faces 3,3,3 -dm_plex_box_lower -1.0,-1.0,-1.0 -dm_plex_simplex 0 -dm_plex_hash_location true -num_comp 2 -uniform_swarm
+
+/// @file
+/// libCEED example using PETSc with DMSwarm
+const char help[] = "libCEED example using PETSc with DMSwarm\n";
+
+#include <ceed.h>
+#include <math.h>
+#include <petscdmplex.h>
+#include <petscdmswarm.h>
+#include <petscds.h>
+#include <petscfe.h>
+#include <petscksp.h>
+#include <petsc/private/petscfeimpl.h> /* For interpolation */
+
+#include "include/petscutils.h"
+
+typedef PetscErrorCode (*DMFunc)(PetscInt dim, PetscReal time, const PetscReal x[], PetscInt Nc, PetscScalar *u, void *ctx);
+
+typedef struct DMSwarmCeedContext_ *DMSwarmCeedContext;
+struct DMSwarmCeedContext_ {
+  Ceed                ceed;
+  CeedVector          u_mesh_loc, v_mesh_loc, u_mesh_elem, u_points_loc, u_points_elem, x_ref_points_loc, x_ref_points_elem;
+  CeedElemRestriction restriction_u_mesh, restriction_x_points, restriction_u_points;
+  CeedBasis           basis_u;
+};
```

Comment:
I think we should revisit this struct when we integrate with Ratel -- parts of this should be stored on the whole mesh, but other parts are field-specific. Maybe something like this for handling multiple fields? Not sure how we will want to minimize redundant work computing particles in cells. 
```c
struct DMSwarmCeedFieldContext_ {
  CeedElemRestriction restriction_u_mesh, restriction_u_points;
  CeedBasis           basis_u;
  const char          fieldname[MAX_NAME_LENGTH];
}
```

Reply:
Definitely - I think we'll ultimately store this into an operator field so we don't have to create a special struct